ql/instruments/creditdefaultswap.cpp

```diff
@@ -101,53 +69,102 @@ namespace QuantLib {
                                          const Date& upfrontDate,
                                          const ext::shared_ptr<Claim>& claim,
                                          const DayCounter& lastPeriodDayCounter,
-                                         const bool rebatesAccrual)
+                                         const bool rebatesAccrual,
+                                         const Date& tradeDate,
+                                         Natural cashSettlementDays)
     : side_(side), notional_(notional), upfront_(upfront),
       runningSpread_(runningSpread), settlesAccrual_(settlesAccrual),
       paysAtDefaultTime_(paysAtDefaultTime), claim_(claim),
-      protectionStart_(protectionStart == Null<Date>() ? schedule[0] :
-                                                         protectionStart) {
-        QL_REQUIRE((protectionStart_ <= schedule[0])
-            || (schedule.rule() == DateGeneration::CDS)
-            || (schedule.rule() == DateGeneration::CDS2015),
-                   "protection can not start after accrual");
+      protectionStart_(protectionStart == Null<Date>() ? schedule[0] : protectionStart),
+      tradeDate_(tradeDate), cashSettlementDays_(cashSettlementDays) {
+
+        init(schedule, convention, dayCounter, lastPeriodDayCounter, rebatesAccrual, upfrontDate);
+    }
+
+    void CreditDefaultSwap::init(const Schedule& schedule, BusinessDayConvention paymentConvention,
+        const DayCounter& dayCounter, const DayCounter& lastPeriodDayCounter,
+        bool rebatesAccrual, const Date& upfrontDate) {
+
+        QL_REQUIRE(!schedule.empty(), "CreditDefaultSwap needs a non-empty schedule.");
+
+        bool postBigBang = false;
+        if (schedule.hasRule()) {
+            DateGeneration::Rule rule = schedule.rule();
+            postBigBang = rule == DateGeneration::CDS || rule == DateGeneration::CDS2015;
```

Comment:
The whole point of introducing the CDS2015 rule was to introduce a rule for post BigBang. This seems strange.